File: Sources/ATTNVisitorService.m
//
//  ATTNVisitorService.m
//  Example
//
//  Created by Olivia Kim on 11/23/22.
//


#import <Foundation/Foundation.h>
#import "ATTNVisitorService.h"
#import "ATTNPersistentStorage.h"


static NSString *const VISITOR_ID_KEY = @"visitorId";


@implementation ATTNVisitorService

- (id)init {
  if (self = [super init]) {
    _persistentStorage = [[ATTNPersistentStorage alloc] init];
  }
  return self;
}

- (NSString *)getVisitorId {
  NSString *existingVisitorId = [_persistentStorage readStringForKey:VISITOR_ID_KEY];
  if (existingVisitorId != nil) {
    return existingVisitorId;
  } else {
    return [self createNewVisitorId];
  }
}

- (NSString *)createNewVisitorId {
  NSString *newVisitorId = [self generateVisitorId];
  [_persistentStorage saveObject:newVisitorId forKey:VISITOR_ID_KEY];
  return newVisitorId;
}

- (NSString *)generateVisitorId {
  NSUUID *uuid = [NSUUID UUID];
  NSString *uuidString = [[uuid UUIDString] stringByReplacingOccurrencesOfString:@"-" withString:@""];
  return [uuidString lowercaseString];
}


@end

File: Sources/ATTNVisitorService.swift
//
//  ATTNVisitorService.swift
//  attentive-ios-sdk-framework
//
//  Created by Vladimir - Work on 2024-05-29.
//

import Foundation

struct ATTNVisitorService {
  private enum Constants {
    static var visitorIdKey: String { "visitorId" }
  }

  private let persistentStorage: ATTNPersistentStorageProtocol

  init(persistentStorage: ATTNPersistentStorageProtocol = ATTNPersistentStorage()) {
    self.persistentStorage = persistentStorage
  }

  func getVisitorId() -> String {
    guard let existingVisitorId = persistentStorage.readString(forKey: Constants.visitorIdKey) else {
      return createNewVisitorId()
    }

    Loggers.event.info("Obtained existing visitor id: \(existingVisitorId)")

    return existingVisitorId
  }

  func createNewVisitorId() -> String {
    let newVisitorId = generateVisitorId()
    persistentStorage.save(newVisitorId as NSObject, forKey: Constants.visitorIdKey)

    Loggers.event.info("Generated new visitor id: \(newVisitorId)")

    return newVisitorId
  }

}

fileprivate extension ATTNVisitorService {
  func generateVisitorId() -> String {
    UUID()
      .uuidString
      .replacingOccurrences(of: "-", with: "")
  }
}
